### ISSUE TITLE
contextlib.contextmanager correct return type

### ISSUE THREAD

**Issue by timkgh** [microsoft/pyright#459]
(Sorry about asking here, I see there's a Gitter channel now)

What is the correct return type in something like this?
```
@contextlib.contextmanager
def f() -> Iterator[Path]:
    ...
    yield path
    ...
```
If I do it as above, at the call site the type seems to be lost for `p`, it shows as unknown in VSCode.
```
with f() as p:
    ...
```
My workaround is to insert a `p = cast(Path, p)` but that is ugly.

I also tried using the `ContextManager` type, as in:
```
@contextlib.contextmanager
def f() -> ContextManager[Path]:
    ...
```
But now there are type errors in the definition.

Is there a good example how to correctly do this?

**Comment by erictraut** [comment]
Thanks for the bug report. This is a bug in pyright's handling of generic protocols. I'll work on a fix.

**Comment by erictraut** [comment]
I have a fix in place with updated tests to cover this condition. The fix will be in the next version of pyright. Thanks again for reporting the problem!

**Comment by erictraut** [comment]
This fix is included in Pyright v1.1.17, which I just published.
